File: perf/lora/gnuradio-rx.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import signal
import sys
import time
from argparse import ArgumentParser
from contextlib import contextmanager

import numpy as np
from gnuradio import blocks
from gnuradio import gr
from gnuradio import lora_sdr

# Parameters matching the compatible FutureSDR file receiver.
BRANCHES = 4
SF = 7
SAMPLE_RATE = 1_000_000
BANDWIDTH = 125_000
OVERSAMPLE = SAMPLE_RATE // BANDWIDTH
CENTER_FREQ = 869_525_000
SYNC_WORD = [0x34]  # public LoRa sync word
SOFT_DECODING = False
LDRO_MODE = False
PREAMBLE_LEN = 8
PRINT_HEADER = False
PRINT_RX_MSG = False

DEFAULT_FILE = "lora-dumps/lora_single_channel_1M_multi_sf_16B_1dBpSF_300s.cf32"


def load_cf32(file):
    dtype = np.dtype("<c8")
    size = os.path.getsize(file)
    if size % dtype.itemsize != 0:
        raise ValueError(
            f"invalid cf32 file size ({size}), expected multiple of {dtype.itemsize} bytes"
        )
    return np.fromfile(file, dtype=dtype)


def fileno(file_or_fd):
    fd = getattr(file_or_fd, "fileno", lambda: file_or_fd)()
    if not isinstance(fd, int):
        raise ValueError("Expected a file (`.fileno()`) or a file descriptor")
    return fd


@contextmanager
def stdout_redirected(to=os.devnull, stdout=None):
    if stdout is None:
        stdout = sys.stdout

    stdout_fd = fileno(stdout)
    with os.fdopen(os.dup(stdout_fd), "wb") as copied:
        stdout.flush()
        with open(to, "wb") as to_file:
            os.dup2(to_file.fileno(), stdout_fd)
        try:
            yield stdout
        finally:
            stdout.flush()
            os.dup2(copied.fileno(), stdout_fd)


class lora_RX(gr.top_block):
    def __init__(self, file=DEFAULT_FILE):
        gr.top_block.__init__(self, "Lora Rx", catch_exceptions=True)

        self.file = file
        self.soft_decoding = SOFT_DECODING
        self.pay_len = 11
        self.impl_head = False
        self.has_crc = True
        self.cr = 1

        samples = load_cf32(file)
        self.sources = []
        self.message_stores = []

        for _ in range(BRANCHES):
            source = blocks.vector_source_c(samples, False, 1, [])
            self.sources.append(source)
            header_decoder = lora_sdr.header_decoder(
                self.impl_head, self.cr, self.pay_len, self.has_crc, LDRO_MODE, PRINT_HEADER
            )
            hamming_dec = lora_sdr.hamming_dec(self.soft_decoding)
            gray_mapping = lora_sdr.gray_mapping(self.soft_decoding)
            frame_sync = lora_sdr.frame_sync(
                int(CENTER_FREQ),
                BANDWIDTH,
                SF,
                self.impl_head,
                SYNC_WORD,
                OVERSAMPLE,
                PREAMBLE_LEN,
            )
            fft_demod = lora_sdr.fft_demod(self.soft_decoding, True)
            dewhitening = lora_sdr.dewhitening()
            deinterleaver = lora_sdr.deinterleaver(self.soft_decoding)
            crc_verif = lora_sdr.crc_verif(PRINT_RX_MSG, False)
            crc_ok_sink = blocks.null_sink(gr.sizeof_char)
            crc_bad_sink = blocks.null_sink(gr.sizeof_char)
            message_debug = blocks.message_debug()
            self.message_stores.append(message_debug)

            self.msg_connect((header_decoder, "frame_info"), (frame_sync, "frame_info"))
            self.msg_connect((crc_verif, "msg"), (message_debug, "store"))
            self.connect((source, 0), (frame_sync, 0))
            self.connect((frame_sync, 0), (fft_demod, 0))
            self.connect((fft_demod, 0), (gray_mapping, 0))
            self.connect((gray_mapping, 0), (deinterleaver, 0))
            self.connect((deinterleaver, 0), (hamming_dec, 0))
            self.connect((hamming_dec, 0), (header_decoder, 0))
            self.connect((header_decoder, 0), (dewhitening, 0))
            self.connect((dewhitening, 0), (crc_verif, 0))
            self.connect((crc_verif, 0), (crc_ok_sink, 0))
            self.connect((crc_verif, 1), (crc_bad_sink, 0))


def count_messages(message_debug):
    count = 0
    while True:
        try:
            message_debug.get_message(count)
            count += 1
        except Exception:
            return count


def argument_parser():
    parser = ArgumentParser()
    parser.add_argument(
        "--run", dest="run", type=int, default=0, help="Set run number [default=%(default)r]"
    )
    parser.add_argument(
        "-f",
        "--file",
        dest="file",
        type=str,
        default=DEFAULT_FILE,
        help="Set file name [default=%(default)r]",
    )
    return parser


def main(top_block_cls=lora_RX, options=None):
    if options is None:
        options = argument_parser().parse_args()

    with stdout_redirected():
        tb = top_block_cls(file=options.file)

        def sig_handler(sig=None, frame=None):
            tb.stop()
            tb.wait()
            sys.exit(0)

        signal.signal(signal.SIGINT, sig_handler)
        signal.signal(signal.SIGTERM, sig_handler)

        start_time = time.time()
        tb.start()
        tb.wait()
        elapsed = time.time() - start_time

    counts = [count_messages(message_store) for message_store in tb.message_stores]
    print(
        "{},{},legacy,{},{},{},{},{}".format(
            options.run, options.file, elapsed, counts[0], counts[1], counts[2], counts[3]
        )
    )


if __name__ == "__main__":
    main()

File: perf/lora/gnuradio-rx.cc
#include <gnuradio/blocks/message_debug.h>
#include <gnuradio/blocks/null_sink.h>
#include <gnuradio/blocks/vector_source.h>
#include <gnuradio/lora_sdr/crc_verif.h>
#include <gnuradio/lora_sdr/deinterleaver.h>
#include <gnuradio/lora_sdr/dewhitening.h>
#include <gnuradio/lora_sdr/fft_demod.h>
#include <gnuradio/lora_sdr/frame_sync.h>
#include <gnuradio/lora_sdr/gray_mapping.h>
#include <gnuradio/lora_sdr/hamming_dec.h>
#include <gnuradio/lora_sdr/header_decoder.h>
#include <gnuradio/top_block.h>

#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <fcntl.h>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <string_view>
#include <unistd.h>
#include <utility>
#include <vector>

namespace {

constexpr std::size_t BRANCHES = 4;
constexpr std::uint8_t SF = 7;
constexpr std::uint32_t SAMPLE_RATE = 1'000'000;
constexpr std::uint32_t BANDWIDTH = 125'000;
constexpr std::uint8_t OVERSAMPLING = SAMPLE_RATE / BANDWIDTH;
constexpr std::uint32_t CENTER_FREQUENCY = 869'525'000;
constexpr std::uint16_t PREAMBLE_LENGTH = 8;
constexpr char DEFAULT_FILE[] =
    "lora-dumps/samples_sf7_pad16_snr30_16B_300s.cf32";

struct Options {
    std::size_t run = 0;
    std::string file = DEFAULT_FILE;
};

class StdoutSilencer {
public:
    StdoutSilencer()
    {
        std::fflush(stdout);
        saved_stdout_ = dup(STDOUT_FILENO);
        const int null_fd = open("/dev/null", O_WRONLY);
        if (saved_stdout_ == -1 || null_fd == -1 || dup2(null_fd, STDOUT_FILENO) == -1) {
            if (null_fd != -1) {
                close(null_fd);
            }
            throw std::runtime_error("could not redirect stdout");
        }
        close(null_fd);
    }

    ~StdoutSilencer()
    {
        std::fflush(stdout);
        if (saved_stdout_ != -1) {
            dup2(saved_stdout_, STDOUT_FILENO);
            close(saved_stdout_);
        }
    }

    StdoutSilencer(const StdoutSilencer&) = delete;
    StdoutSilencer& operator=(const StdoutSilencer&) = delete;

private:
    int saved_stdout_ = -1;
};

struct Flowgraph {
    gr::top_block_sptr top_block;
    std::vector<gr::blocks::message_debug::sptr> message_stores;
};

std::string option_value(int& index, int argc, char** argv, std::string_view name)
{
    const std::string_view argument(argv[index]);
    const std::string prefix = std::string(name) + "=";
    if (argument.starts_with(prefix)) {
        return std::string(argument.substr(prefix.size()));
    }
    if (argument == name && index + 1 < argc) {
        return argv[++index];
    }
    return {};
}

Options parse_options(int argc, char** argv)
{
    Options options;
    for (int index = 1; index < argc; ++index) {
        if (const auto value = option_value(index, argc, argv, "--run"); !value.empty()) {
            options.run = std::stoul(value);
        } else if (const auto value = option_value(index, argc, argv, "--file");
                   !value.empty()) {
            options.file = value;
        } else if (std::string_view(argv[index]) == "-f" && index + 1 < argc) {
            options.file = argv[++index];
        } else {
            throw std::runtime_error("unknown or incomplete argument: " +
                                     std::string(argv[index]));
        }
    }
    return options;
}

std::vector<gr_complex> load_cf32(const std::string& path)
{
    static_assert(sizeof(gr_complex) == 8);

    std::ifstream input(path, std::ios::binary | std::ios::ate);
    if (!input) {
        throw std::runtime_error("could not open IQ dump: " + path);
    }
    const auto size = input.tellg();
    if (size < 0 || size % static_cast<std::streamoff>(sizeof(gr_complex)) != 0) {
        throw std::runtime_error("invalid CF32 IQ dump size: " + path);
    }

    std::vector<gr_complex> samples(static_cast<std::size_t>(size) / sizeof(gr_complex));
    input.seekg(0);
    input.read(reinterpret_cast<char*>(samples.data()), size);
    if (!input) {
        throw std::runtime_error("could not read complete IQ dump: " + path);
    }
    return samples;
}

Flowgraph build_flowgraph(const std::vector<gr_complex>& samples)
{
    auto top_block = gr::make_top_block("LoRa RX");
    std::vector<gr::blocks::message_debug::sptr> message_stores;
    message_stores.reserve(BRANCHES);

    for (std::size_t branch = 0; branch < BRANCHES; ++branch) {
        auto source = gr::blocks::vector_source_c::make(samples, false);
        auto frame_sync = gr::lora_sdr::frame_sync::make(CENTER_FREQUENCY,
                                                         BANDWIDTH,
                                                         SF,
                                                         false,
                                                         { 0x34 },
                                                         OVERSAMPLING,
                                                         PREAMBLE_LENGTH);
        auto fft_demod = gr::lora_sdr::fft_demod::make(false, true);
        auto gray_mapping = gr::lora_sdr::gray_mapping::make(false);
        auto deinterleaver = gr::lora_sdr::deinterleaver::make(false);
        auto hamming_dec = gr::lora_sdr::hamming_dec::make(false);
        auto header_decoder =
            gr::lora_sdr::header_decoder::make(false, 1, 11, true, 0, false);
        auto dewhitening = gr::lora_sdr::dewhitening::make();
        auto crc_verif = gr::lora_sdr::crc_verif::make(0, false);
        auto crc_ok_sink = gr::blocks::null_sink::make(sizeof(std::uint8_t));
        auto crc_bad_sink = gr::blocks::null_sink::make(sizeof(std::uint8_t));
        auto message_store = gr::blocks::message_debug::make();

        top_block->msg_connect(header_decoder, "frame_info", frame_sync, "frame_info");
        top_block->msg_connect(crc_verif, "msg", message_store, "store");
        top_block->connect(source, 0, frame_sync, 0);
        top_block->connect(frame_sync, 0, fft_demod, 0);
        top_block->connect(fft_demod, 0, gray_mapping, 0);
        top_block->connect(gray_mapping, 0, deinterleaver, 0);
        top_block->connect(deinterleaver, 0, hamming_dec, 0);
        top_block->connect(hamming_dec, 0, header_decoder, 0);
        top_block->connect(header_decoder, 0, dewhitening, 0);
        top_block->connect(dewhitening, 0, crc_verif, 0);
        top_block->connect(crc_verif, 0, crc_ok_sink, 0);
        top_block->connect(crc_verif, 1, crc_bad_sink, 0);
        message_stores.push_back(std::move(message_store));
    }

    return { std::move(top_block), std::move(message_stores) };
}

} // namespace

int main(int argc, char** argv)
{
    try {
        const auto options = parse_options(argc, argv);
        auto samples = load_cf32(options.file);
        Flowgraph flowgraph;
        double elapsed_seconds;

        {
            StdoutSilencer silence;
            flowgraph = build_flowgraph(samples);
            std::vector<gr_complex>().swap(samples);

            const auto start = std::chrono::steady_clock::now();
            flowgraph.top_block->start();
            flowgraph.top_block->wait();
            elapsed_seconds =
                std::chrono::duration<double>(std::chrono::steady_clock::now() - start).count();
        }

        std::cout << options.run << ',' << options.file << ",legacy," << elapsed_seconds;
        for (const auto& message_store : flowgraph.message_stores) {
            std::cout << ',' << message_store->num_messages();
        }
        std::cout << '\n';
    } catch (const std::exception& error) {
        std::cerr << "gnuradio-rx: " << error.what() << '\n';
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}
